### ISSUE TITLE
Python 37, 38, 39: Tensorflow broken

### ISSUE THREAD

**Issue by bergkvist** [NixOS/nixpkgs#128829]
**Describe the bug**
When trying to add a LSTM-layer to a tensorflow model, I get the following error:
```
NotImplementedError: Cannot convert a symbolic Tensor (lstm/strided_slice:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported
```
It seems like Tensorflow is not compatible with the version of Numpy included in nixpkgs.

Relevant links:
- https://stackoverflow.com/questions/66207609/notimplementederror-cannot-convert-a-symbolic-tensor-lstm-2-strided-slice0-t/66207610
- https://stackoverflow.com/questions/58479556/notimplementederror-cannot-convert-a-symbolic-tensor-2nd-target0-to-a-numpy
- https://github.com/tensorflow/models/issues/9706
- https://github.com/tensorflow/tensorflow/issues/47691

**To Reproduce**
```py
# example.py
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM
model = Sequential()
model.add(LSTM(256, input_shape=(1, 66), return_sequences=True))
```
```sh
# The same issue can be reproduced with python37, python38, python39. I didn't test older versions.
nix-shell -p "python37.withPackages(ps: [ps.tensorflow])" --run "python example.py" --pure
nix-shell -p "python38.withPackages(ps: [ps.tensorflow])" --run "python example.py" --pure
nix-shell -p "python39.withPackages(ps: [ps.tensorflow])" --run "python example.py" --pure
```

**Expected behavior**
Adding the LSTM-layer should not throw an error.

**Traceback**
```
2021-06-30 23:51:17.280747: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Traceback (most recent call last):
  File "/home/tobias/example.py", line 6, in <module>
    model.add(layer)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/training/tracking/base.py", line 517, in _method_wrapper
    result = method(self, *args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/sequential.py", line 208, in add
    layer(x)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 660, in __call__
    return super(RNN, self).__call__(inputs, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 951, in __call__
    return self._functional_construction_call(inputs, args, kwargs,
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 1090, in _functional_construction_call
    outputs = self._keras_tensor_symbolic_call(
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 822, in _keras_tensor_symbolic_call
    return self._infer_output_signature(inputs, args, kwargs, input_masks)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 863, in _infer_output_signature
    outputs = call_fn(inputs, *args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent_v2.py", line 1157, in call
    inputs, initial_state, _ = self._process_inputs(inputs, initial_state, None)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 859, in _process_inputs
    initial_state = self.get_initial_state(inputs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 642, in get_initial_state
    init_state = get_initial_state_fn(
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 2506, in get_initial_state
    return list(_generate_zero_filled_state_for_cell(
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 2987, in _generate_zero_filled_state_for_cell
    return _generate_zero_filled_state(batch_size, cell.state_size, dtype)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 3003, in _generate_zero_filled_state
    return nest.map_structure(create_zeros, state_size)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/util/nest.py", line 659, in map_structure
    structure[0], [func(*x) for x in entries],
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/util/nest.py", line 659, in <listcomp>
    structure[0], [func(*x) for x in entries],
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 3000, in create_zeros
    return array_ops.zeros(init_state_size, dtype=dtype)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py", line 2819, in wrapped
    tensor = fun(*args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py", line 2868, in zeros
    output = _constant_if_small(zero, shape, dtype, name)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py", line 2804, in _constant_if_small
    if np.prod(shape) < 1000:
  File "<__array_function__ internals>", line 5, in prod
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 3030, in prod
    return _wrapreduction(a, np.multiply, 'prod', axis, dtype, out,
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 87, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/framework/ops.py", line 852, in __array__
    raise NotImplementedError(
NotImplementedError: Cannot convert a symbolic Tensor (lstm/strided_slice:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported
```

**Additional context**

**Notify maintainers**

@jyp
@abbradar 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.41-1-MANJARO, Manjaro Linux, noversion`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.12`
 - channels(tobias): `"nixpkgs-21.11pre294396.3bc8e5cd23b, nixpkgs-afdb5675a18-21.11pre293818.afdb5675a18, pkgs-21.11pre294396.3bc8e5cd23b, pkgs-afdb5675a18-21.11pre293818.afdb5675a18"`
 - nixpkgs: `/home/tobias/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by CMCDragonkai** [comment]
Would this be fixed if we upgraded to TF 2.5. There are other breakages associated with not having 2.5 https://github.com/NixOS/nixpkgs/pull/125328

**Comment by bergkvist** [comment]
It might not be possible to get Tensorflow working with nixpkgs without downgrading `pkgs.python(37,38,39).pkgs.numpy` to 1.19 (from 1.20): https://github.com/tensorflow/tensorflow/issues/47691. Other libraries that depend on numpy 1.20 might break as a result.

Python doesn't allow importing 2 different libraries with the same name. So it makes little sense to provide `pkgs.python.pkgs.numpy_1_19` in addition to `pkgs.python.pkgs.numpy_1_20`, since this will break any environment that depends on both versions. You lose the ability to compose Python packages in a sane way.

A workaround for being able to import the same library twice might be to patch numpy 1.19 and rename it to something like numpy_1_19.
```py
import numpy_1_19 as np
```

Tensorflow would also need to be patched to import numpy_1_19 instead of numpy.

Whether this patching could be done/automated in a sane way is a different question.

**Comment by alexvorobiev** [comment]
There is a patch for that particular error https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-875132015. I haven't tested it.

**Comment by bergkvist** [comment]
That patch seems to make my test above pass. I'll submit a PR for this today.

**Comment by alexvorobiev** [comment]
There is now this PR upstream https://github.com/tensorflow/tensorflow/pull/48935.

**Comment by risicle** [comment]
Fixed in #132099

**Comment by alexvorobiev** [comment]
According to https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-944003768 2.6 supports the new numpy.
